### PR TITLE
Tests/fix for importing variables/mixins

### LINKED ISSUE #51: Unable to import mixins/variables
Using postcss-advanced-variables 2.1.x, I'm not able to define a mixin or a variable in one file, and then import that file to use the mixin/variable in another file.

Failing tests coming momentarily.

### PR DESCRIPTION
**Pull request by TehShrike**
Failing tests for #51 

I'll start poking to see if I can figure out what's going on.  Let me know if you see any mistakes in the tests.

---

Fixes #51

### CODE REVIEW

File: test/imports/mixins.css [hunk 1]
```diff
@@ -0,0 +1,8 @@
+@import 'imports/mixins';
```
Comment (jonathantneal) [inline test/imports/mixins.css]: I think you mean to import this in your test and not in the file itself.

File: test/import-mixins.css [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.test-1 {
```
Comment (jonathantneal) [inline test/import-mixins.css]: I think you want to import the mixins file here

File: test/imports/mixins.css (hunk 1 above)
Comment (TehShrike) [reply test/imports/mixins.css]: d'oh!

**Comment by TehShrike** [comment]
Fixed the silly `@import` placement.  It still throws the same `Could not resolve the mixin for "mixin-test-1"` message.

**Comment by jonathantneal** [comment]
The issue here is that the `@mixin` is evaluated before the promise of the `@import` has finished.

**Comment by TehShrike** [comment]
So I'm new to this code base, but my first instinct is to make `transformNode` smarter so that if any of the transform functions return a promise, wait until the promise is resolved before continuing to transform further nodes.

Does that make sense?  Did I understand your diagnosis correctly?

Should I implement that, or are you leaning a different direction?

**Comment by TehShrike** [comment]
Ah, I see `importPromise` is passed around.  I suppose it would be easier to just change `transform-content-atrule`/`transform-include-atrule`/`get-replaced-string` to wait on `importPromise` before proceeding.

I feel like it might be preferable to make all the transforms async instead of making them have to know that they could be affected by other imports.

That would be a bigger change though, so I'll update this pull request with the easier `importPromise` implementation for now, unless you say otherwise.

**Comment by TehShrike** [comment]
Oh nope - `importPromise` is currently mutated by `transformImportAtrule`, so I think "wait until the promises in `importPromise` are all resolved" would be unreliable in this case:

```sass
/* root.css */
@import 'child.css';

.foo {
	content: $bar;
}
```

```sass
/* child.css */
@import 'grandchild.css';
```

```sass
/* grandchild.css */
$bar: 'sup dawg':
```

Because (if I'm reading the code correctly), when `$bar` would be evaluated, only the promise for child.css would be in `importPromise`.  The `grandchild.css` promise wouldn't be added to that array until after `child.css` was loaded and processed.

So assuming I haven't missed something, I think making `transformNode` and the specific transform functions properly async and handling everything in order may be necessary.

Let me know if I'm on track and I can move ahead with implementation :-)

**Comment by jonathantneal** [comment]
You are correct. Except with Imports we don’t know how the rule chain will be impacted, so it’s best we just give resolutions a “second chance” after importPromise has fully resolved.

**Comment by TehShrike** [comment]
Well, I took a swing at making async transformation a thing.  It feels pretty close, but there's still an issue I can't figure out in `@each` transformation.

Let me know if you see where I went wrong.  If you want to pair code at some point in the next few days, that would be cool too.

Let me know if you have any issues with the direction of the refactor, too.

**Comment by TehShrike** [comment]
Got it.

`setVariable` was just mutating the variables object on the parent rule, and every clone was getting its variables object set to that shared object.

Shallow copying the variables object before putting it onto the clone fixes that issue - the tests are all passing now.

---

A couple of the conditional Promise parts aren't as pretty/readable as I'd like (mostly in `transform-if-atrule`).  async/await would make it a lot cleaner, but support for node 4 doesn't drop until the end of April, and adding the necessary babel runtime to the build didn't seem appropriate.

I feel like this is generally the correct approach - allow transformations to be async, and don't force transforms to watch any global promises/mutexes.

Let me know what you think!

**Comment by jonathantneal** [comment]
Brilliant. I tried a few things and could never crack it.

FWIW, Babel is already in the project. https://github.com/jonathantneal/postcss-advanced-variables/blob/master/.rollup.js

**Comment by TehShrike** [comment]
I did try adding in async/await, but babel adds so much extra code, and then it was complaining that I needed to import more babel runtime files, so I just went back to `.then()` :-x

If you're cool with dropping node 4 support once its maintenance window ends, we can always pretty it up in a few months.

**Pull request merged**